File: internal/parser/list_items.go
package parser

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/noclaps/znak/highlight"
)

type listType int

const (
	orderedList listType = iota
	unorderedList
)

func listItems(input string, codeTheme highlight.Theme, listType listType) ([]node, error) {
	var re *regexp.Regexp
	var segMatch string
	switch listType {
	case orderedList:
		re = regexp.MustCompile(`(?m)^\d+\. `)
		segMatch = "   "
	case unorderedList:
		re = regexp.MustCompile(`(?m)^- `)
		segMatch = "  "
	}

	reSplitInput := re.Split(input, -1)
	lines := []string{}
	for _, line := range reSplitInput {
		if line == "" {
			continue
		}
		lines = append(lines, strings.TrimSpace(line))
	}

	tokens := []node{}
	for _, line := range lines {
		segments := strings.Split(line, "\n")
		if len(segments) == 1 {
			children, err := Parse(line, codeTheme)
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, element{TagName: "li", Children: children})
		} else {
			remainingLines := ""
			for _, line := range segments[1:] {
				remainingLines += fmt.Sprintln(strings.Replace(line, segMatch, "", 1))
			}
			input := fmt.Sprintf("%s\n\n%s", segments[0], remainingLines)
			children, err := Parse(input, codeTheme)
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, element{TagName: "li", Children: children})
		}
	}

	return tokens, nil
}

File: crates/znak/src/parser/list_items.rs
use std::mem;

use highlight::Highlight;

use crate::parser::{
    parse,
    types::{Node, element},
};

pub(crate) enum ListType {
    Ordered,
    Unordered,
}

pub(crate) fn list_items(input: String, hl: &Highlight, list_type: ListType) -> Vec<Node> {
    let (split_input, seg_match) = match list_type {
        ListType::Ordered => {
            let input = input.lines();
            let mut lines = vec![];
            let mut val = String::new();
            for line in input {
                let mut chars = line.chars().skip_while(|c| c.is_digit(10));
                if chars.next().is_some_and(|c| c == '.') && chars.next().is_some_and(|c| c == ' ')
                {
                    lines.push(mem::take(&mut val));
                    val += chars.collect::<String>().as_str();
                    val += "\n";
                } else {
                    val += line;
                    val += "\n";
                };
            }
            lines.push(mem::take(&mut val));
            (lines[1..].to_vec(), "   ")
        }
        ListType::Unordered => {
            let input = input.lines();
            let mut lines = vec![];
            let mut val = String::new();
            for line in input {
                if line.starts_with("- ") {
                    lines.push(mem::take(&mut val));
                    val += &line[2..];
                    val += "\n";
                } else {
                    val += line;
                    val += "\n";
                }
            }
            lines.push(mem::take(&mut val));
            (lines, "  ")
        }
    };

    let mut lines = vec![];
    for line in split_input.iter().map(|l| l.as_str()) {
        if line == "" {
            continue;
        }
        lines.push(line.trim());
    }

    let mut tokens = vec![];
    for line in lines {
        let segments = line.split("\n").collect::<Vec<&str>>();
        if segments.len() == 1 {
            let children = parse(segments[0].to_string(), hl);
            tokens.push(element!("li", children));
        } else {
            let mut remaining_lines = String::new();
            for &line in segments[1..].iter() {
                remaining_lines += format!("{}\n", line.replacen(seg_match, "", 1)).as_str();
            }
            let input = format!("{}\n\n{}", segments[0], remaining_lines);
            let children = parse(input, hl);
            tokens.push(element!("li", children));
        }
    }

    tokens
}
